github/retry.go

```diff
@@ -1,90 +1,93 @@
 package github
 
 import (
-	"bytes"
 	"context"
 	"fmt"
-	"io/ioutil"
 	"net/http"
 	"time"
 
+	"github.com/src-d/metadata-retrieval/utils/httpbody"
+
+	"github.com/cenkalti/backoff"
 	"gopkg.in/src-d/go-log.v1"
 )
 
-// errUnretriable wraps an error to stop retry
-type errUnretriable struct {
-	Err error
-}
-
-func (e *errUnretriable) Error() string {
-	return e.Err.Error()
+// SetRetryTransport wraps the passed client.Transport with a RetryTransport
+func SetRetryTransport(client *http.Client) {
+	client.Transport = &retryTransport{client.Transport}
 }
 
+// retryTransport retries a http.Request if it fails when processing, or if
+// its http.Response has StatusCode in 5xx range (server errors)
 type retryTransport struct {
 	T http.RoundTripper
 }
 
 func (t *retryTransport) RoundTrip(req *http.Request) (*http.Response, error) {
-	var r *http.Response
-	var err error
-	retry(func() error {
-		r, err = t.T.RoundTrip(req)
-		if err != nil {
-			if err == context.Canceled {
-				return &errUnretriable{Err: err}
-			}
+	var response *http.Response
+	backupReqBody := req.Body
+	do := func() error {
+		var err error
 
-			return err
+		req.Body = backupReqBody
+		backupReqBody, err = httpbody.BackupRequestBody(req)
+		if err != nil {
+			return fmt.Errorf("could not backup the response before sending it through the retry loop: %s", err)
 		}
 
-		if r.StatusCode == http.StatusOK {
-			return nil
+		response, err = t.T.RoundTrip(req)
+		if err == context.Canceled {
+			return backoff.Permanent(err)
 		}
 
-		body, _ := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return err
+		}
 
-		// Restore the io.ReadCloser
-		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		if response.StatusCode >= 500 {
+			responseBody, err := httpbody.ReadResponseAndRestore(response)
+			if err != nil {
+				return fmt.Errorf("could not backup the response before logging the HTTP %d response: %s", response.StatusCode, err)
+			}
 
-		err = fmt.Errorf("non-200 OK status code: %v body: %q", r.Status, body)
-		if r.StatusCode > 500 {
-			return err
+			return fmt.Errorf("%s: %s", response.Status, responseBody)
 		}
-		return &errUnretriable{Err: err}
-	})
 
-	return r, err
+		return nil
```

Comment:
why you don't return error for client error anymore?

Reply:
it's already done by `shurcooL/graphql` and imo it's not the responsibility of a `RetryTransport` to parse responses that should not be retried, and then send errors.
That should be done by a new and hypothetical `ErrorHandlerTransport`, for example (but I'm not sure if it will even go against `Transport` interface...) If so, it could be great.

Reply:
> it's already done by shurcooL/graphql

👍